File: app/src/androidTest/java/ray/easydev/fragmentnav/sample/BaseFmTest.java
package ray.easydev.fragmentnav.sample;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.filters.LargeTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Rule;
import org.junit.runner.RunWith;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import ray.easydev.fragmentnav.FnActivity;
import ray.easydev.fragmentnav.FragmentNav;

/**
 * Created by Ray on 2017/12/1.
 */

@RunWith(AndroidJUnit4.class)
@LargeTest
public class BaseFmTest {
    final static Class TAG = BaseFmTest.class;

    @Rule
    public ActivityTestRule<MainActivity> mActivityRule = new ActivityTestRule<>(
            MainActivity.class);

    private FragmentActivity mActivity;
    private FragmentManager fragmentManager;
    private Context mContext;

    FnFragments fnFragments;
    SysFragments sysFragments;

    @Before
    public void setup(){
        mContext = InstrumentationRegistry.getTargetContext();
        mActivity = mActivityRule.getActivity();
        fragmentManager = mActivity.getSupportFragmentManager();

        fnFragments = new FnFragments(((FnActivity) mActivity).getFragmentNav());
        sysFragments = new SysFragments(fragmentManager);
    }

    public FragmentActivity getActivity(){
        return mActivity;
    }

    public Context getContext(){
        return mContext;
    }

    public FnFragments getFnFragments() {
        return fnFragments;
    }

    public SysFragments getSysFragments() {
        return sysFragments;
    }

    public static String message(String format, Object... args){
        return String.format(format, args);
    }

    static class FnFragments {
        FragmentNav fragmentNav;

        FnFragments(FragmentNav fragmentNav){
            this.fragmentNav = fragmentNav;
        }

        Fragment getCurrentFragment(){
            return fragmentNav.getCurrentFragment();
        }

        List<Fragment> getFragments(){
            List<Fragment> fragments = new ArrayList<>();
            Collection<Integer> ids = fragmentNav.taskIds();
            for (Integer id : ids) {
                fragments.addAll(fragmentNav.getFragments(id));
            }

            return fragments;
        }
    }

    static class SysFragments {
        FragmentManager fragmentManager;

        SysFragments(FragmentManager fragmentManager){
            this.fragmentManager = fragmentManager;
        }

        Fragment getCurrentFragment(){
            List<Fragment> fragments = getFragments();
            Assert.assertTrue(message("Sys fragments' size is 0"), fragments.size() > 0);
            return fragments.get(fragments.size() - 1);
        }

        List<Fragment> getFragments(){
            try {
                Field field = fragmentManager.getClass().getDeclaredField("mAdded");
                field.setAccessible(true);
                return (List<Fragment>) field.get(fragmentManager);
            } catch (Exception e){

            }

            return new ArrayList<>(0);
        }
    }


}

File: app/src/androidTest/java/ray/easydev/fragmentnav/sample/BaseFmTest.kt
package ray.easydev.fragmentnav.sample

import android.content.Context
import android.support.test.InstrumentationRegistry
import android.support.test.rule.ActivityTestRule
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentManager
import junit.framework.Assert
import org.junit.Before
import org.junit.Rule
import ray.easydev.fragmentnav.BaseFmTest
import ray.easydev.fragmentnav.FnActivity
import ray.easydev.fragmentnav.FnFragment
import ray.easydev.fragmentnav.FragmentIntent
import ray.easydev.fragmentnav.FragmentNav
import java.util.*

/**
 * Created by Ray on 2017/12/1.
 */

open class BaseFmTest {

    @Rule
    @JvmField
    public var mActivityRule = ActivityTestRule(
            MainActivity::class.java)

    lateinit var activity: FragmentActivity
        private set

    lateinit var context: Context
        private set

    lateinit var fragmentManager: FragmentManager
    lateinit var fnFragments: FnFragments
        internal set
    lateinit var sysFragments: SysFragments
        internal set

    @Before
    fun setup() {
        context = InstrumentationRegistry.getTargetContext()
        activity = mActivityRule.activity
        fragmentManager = activity.supportFragmentManager

        fnFragments = FnFragments((activity as FnActivity).fragmentNav)
        sysFragments = SysFragments(fragmentManager)
    }

    internal fun startFragment(vararg intents: FragmentIntent) {
        fnFragments.currentFragment.startFragment(*intents)
    }

    fun finish(){
        fnFragments.currentFragment.finish()
    }

    fun finishTask(){
        fnFragments.currentFragment.finishTask()
    }

    class FnFragments(var fragmentNav: FragmentNav) {

        val currentFragment: FnFragment
            get() = fragmentNav.currentFragment

        val fragments: List<Fragment>
            get() {
                val fragments = ArrayList<Fragment>()
                val ids = fragmentNav.taskIds()
                for (id in ids) {
                    fragments.addAll(fragmentNav.getFragments(id!!))
                }

                return fragments
            }
    }

    class SysFragments(var fragmentManager: FragmentManager) {

        val currentFragment: Fragment
            get() {
                val fragments = fragments
                Assert.assertTrue(message("Sys fragments' size is 0"), fragments.size > 0)
                return fragments[fragments.size - 1]
            }

        val fragments: List<Fragment>
            get() = fragmentManager.fragments
    }

    internal fun checkOrder() {
        if (fnFragments.currentFragment === sysFragments.currentFragment) {
            val fragments0 = fnFragments.fragments
            val fragments1 = sysFragments.fragments

            if (fragments0.size == fragments1.size) {
                for (i in fragments0.indices) {
                    val fragment0 = fragments0[i]
                    val fragment1 = fragments1[i]

                    if (fragment0 != fragment1) {
                        Assert.fail()
                    }
                }

                return;
            }
        }

        return Assert.fail( "Current fragment ")
    }

    internal fun checkTask(taskSize: Int) {
        val fragmentNav = fnFragments.fragmentNav

        val taskIds = fragmentNav.taskIds()
        Assert.assertTrue(fragmentNav.taskIds().size == taskSize)

        val fragments = fragmentNav.getFragments(taskIds[taskIds.size - 1])
        Assert.assertTrue(!fragments!!.isEmpty())
        Assert.assertTrue(fragments.last() === fnFragments.currentFragment && fnFragments.currentFragment === sysFragments.currentFragment)
    }

    fun checkCurrent(cls: Class<out Fragment>) {
        var visibleCount = 0;
        sysFragments.fragments.forEach {
            if(it.isVisible) visibleCount ++
        }

        Assert.assertTrue((visibleCount == 1) && (sysFragments.currentFragment::class.java == cls))
    }

    internal fun checkState(taskSize: Int, currFragment: Class<out Fragment>) {
        checkTask(taskSize)
        checkCurrent(currFragment)
        checkOrder()
    }

    companion object {
        internal val TAG: Class<*> = BaseFmTest::class.java

        fun message(format: String, vararg args: Any): String {
            return String.format(format, *args)
        }
    }

}
